networkx/algorithms/regular.py

```diff
@@ -109,107 +112,57 @@ def k_factor(G, k, matching_weight="weight"):
-                for neighbor, edge_attrs in list(adj_view.items()):
-                    if neighbor not in self.core_vertices:
-                        self.g.add_edge(self.original, neighbor, **edge_attrs)
-                        break
-            g.remove_nodes_from(self.outer_vertices)
-            g.remove_nodes_from(self.core_vertices)
-
-    class SmallKGadget:
-        def __init__(self, k, degree, node, g):
-            self.original = node
-            self.k = k
-            self.degree = degree
-            self.g = g
-
-            self.outer_vertices = [(node, x) for x in range(degree)]
-            self.inner_vertices = [(node, x + degree) for x in range(degree)]
-            self.core_vertices = [(node, x + 2 * degree) for x in range(k)]
-
-        def replace_node(self):
-            adj_view = self.g[self.original]
-            for outer, inner, (neighbor, edge_attrs) in zip(
-                self.outer_vertices, self.inner_vertices, list(adj_view.items())
-            ):
-                self.g.add_edge(outer, inner)
-                self.g.add_edge(outer, neighbor, **edge_attrs)
-            for core in self.core_vertices:
-                for inner in self.inner_vertices:
-                    self.g.add_edge(core, inner)
-            self.g.remove_node(self.original)
-
-        def restore_node(self):
-            self.g.add_node(self.original)
-            for outer in self.outer_vertices:
-                adj_view = self.g[outer]
-                for neighbor, edge_attrs in adj_view.items():
-                    if neighbor not in self.core_vertices:
-                        self.g.add_edge(self.original, neighbor, **edge_attrs)
-                        break
-            self.g.remove_nodes_from(self.outer_vertices)
-            self.g.remove_nodes_from(self.inner_vertices)
-            self.g.remove_nodes_from(self.core_vertices)
-
-    # Step 1
+    # Validate minimum degree requirement.
     if any(d < k for _, d in G.degree):
         raise nx.NetworkXUnfeasible("Graph contains a vertex with degree less than k")
-    g = G.copy()
 
-    # Step 2
+    g = G.copy()
     gadgets = []
-    for node, degree in list(g.degree):
-        if k < degree / 2.0:
-            gadget = SmallKGadget(k, degree, node, g)
+
+    # Replace each node with a gadget.
+    for node, degree in G.degree():
+        is_large = k >= degree / 2.0
+
+        # Create gadget nodes.
+        outer = [(node, i) for i in range(degree)]
+        if is_large:
+            core = [(node, i) for i in range(degree, 2 * degree - k)]
+            inner = []
         else:
-            gadget = LargeKGadget(k, degree, node, g)
-        gadget.replace_node()
-        gadgets.append(gadget)
+            core = [(node, i) for i in range(2 * degree, 2 * degree + k)]
+            inner = [(node, i) for i in range(degree, 2 * degree)]
+
+        # Connect gadget nodes to neighbors.
+        if not is_large:
+            g.add_edges_from(zip(outer, inner))
```

Comment:
If I'm following correctly, I think the conditional is not necessary, as inner will be empty if `is_large` and zipping with one of the iterables empty will just terminate.

That said, I don't think there's anything wrong with being explicit - just double-checking that I'm following the logic correctly!